homeassistant/components/ecobee/config_flow.py

```diff
@@ -0,0 +1,160 @@
+            self._ecobee = Ecobee(config={ECOBEE_API_KEY: user_input[CONF_API_KEY]})
+
+            if await self.hass.async_add_executor_job(self._ecobee.request_pin):
+                """We have a PIN; move to the next step of the flow."""
+                return await self.async_step_authorize()
+            errors["base"] = "pin_request_failed"
+
+        return self.async_show_form(
+            step_id="init",
+            data_schema=vol.Schema(
+                {vol.Required(CONF_API_KEY, default=stored_api_key): str}
+            ),
+            errors=errors,
+        )
+
+    async def async_step_authorize(self, user_input=None):
+        """Present the user with the PIN so that the app can be authorized on ecobee.com."""
+        errors = {}
+
+        if user_input is not None:
+            """Attempt to obtain tokens from ecobee and finish the flow."""
+            if await self.hass.async_add_executor_job(self._ecobee.request_tokens):
+                """Refresh token obtained; create the config entry."""
+                config = {
+                    CONF_API_KEY: self._ecobee.api_key,
+                    CONF_REFRESH_TOKEN: self._ecobee.refresh_token,
+                }
+                return self.async_create_entry(title=DOMAIN, data=config)
+            errors["base"] = "token_request_failed"
+
+        return self.async_show_form(
+            step_id="authorize",
+            errors=errors,
+            description_placeholders={"pin": self._ecobee.pin},
+        )
+
+    async def async_step_import(self, import_data):
+        """
+        Import ecobee config from configuration.yaml.
+
+        Triggered by async_setup only if a config entry doesn't already exist.
+        If ecobee.conf exists, we will attempt to validate the credentials
+        and create an entry if valid. Otherwise, we will delegate to the user
+        step so that the user can continue the config flow.
+        """
+        try:
+            legacy_config = await self.hass.async_add_executor_job(
+                load_json, self.hass.config.path(ECOBEE_CONFIG_FILENAME)
+            )
+            ecobee = Ecobee(
+                config={
+                    ECOBEE_API_KEY: legacy_config[ECOBEE_API_KEY],
+                    ECOBEE_REFRESH_TOKEN: legacy_config[ECOBEE_REFRESH_TOKEN],
+                }
+            )
+            if await self.hass.async_add_executor_job(ecobee.refresh_tokens):
+                """Credentials found and validated; create the entry."""
+                _LOGGER.debug(
+                    "Valid ecobee configuration found for import, creating config entry"
+                )
+                return self.async_create_entry(
+                    title=DOMAIN,
+                    data={
+                        CONF_API_KEY: ecobee.api_key,
+                        CONF_REFRESH_TOKEN: ecobee.refresh_token,
+                    },
+                )
+        except Exception:  # pylint: disable=broad-except
```

Comment:
We should catch one or more specific errors. What error are we expecting?

Please wrap as little as possible in each try... except. Move the except statement up under loading json if that's where the first error can happen.

Reply:
The only real expected exception is a KeyError if legacy_config is {} (which could happen because ecobee.conf doesn't exist or is malformed. But I noticed that load_json can raise other exceptions too. Here I used the broad except because it doesn't really matter for the purposes of the config flow import step _why_ the import failed, only that it did, because then we have to delegate to the user step. And testing if ecobee.refresh_tokens is true should only happen if the ecobee object was created (i.e. no exception had been thrown up to that point).

Reply:
Please catch `KeyError` and `HomeAssistantError` and move the except up under `load_json`.

Reply:
The KeyError would be raised by the following if either ECOBEE_API_KEY or ECOBEE_REFRESH_TOKEN aren't present in the legacy config.
```
ecobee = Ecobee(
                config={
                    ECOBEE_API_KEY: legacy_config[ECOBEE_API_KEY],
                    ECOBEE_REFRESH_TOKEN: legacy_config[ECOBEE_REFRESH_TOKEN],
                }
            )
```
If I move the except under that clause, then the next IF statement will raise an exception if the exception was raised previously since the ecobee object won't exist. I think I should leave the code in place as is but I can except on only KeyError and HomeAssistantError  rather than broad except. Does that work?